### ISSUE TITLE
HTML highlighting conflicts with matchparen

### ISSUE THREAD

**Issue by baltitenger** [nvim-treesitter/nvim-treesitter#2603]
### Describe the bug

HTML highlighting gets lost when moving the cursor over anything in `matchpairs` inside a tag more than 5 levels deep. `:TSHighlightCapturesUnderCursor` reports `No highlight groups found` inside these tags, even when they are highlighted correctly. Running `:NoMatchParen` apparently fixes this.

Demo: https://asciinema.org/a/VtBHvpyN15hLIzCyoGCQfBEos

### To Reproduce

1. open an html document
2. move the cursor to the opening `<` of a tag that's more than 5 levels deep
3. see that the whole tag loses highlighting

### Expected behavior

The highlights should not change when moving the cursor around.

### Output of `:checkhealth nvim-treesitter`

```markdown
nvim-treesitter: require("nvim-treesitter.health").check()
========================================================================
## Installation
  - OK: `tree-sitter` found 0.20.5 (parser generator, only needed for :TSInstallFromGrammar)
  - OK: `node` found v17.6.0 (only needed for :TSInstallFromGrammar)
  - OK: `git` executable found.
  - OK: `cc` executable found. Selected from { vim.NIL, "cc", "gcc", "clang", "cl", "zig" }
    Version: cc (GCC) 11.2.0
  - OK: Neovim was compiled with tree-sitter runtime ABI version 14 (required >=13). Parsers must be compatible with runtime ABI.

## Parser/Features H L F I J
  - css            ✓ . ✓ ✓ ✓ 
  - html           ✓ ✓ ✓ ✓ ✓ 
  - javascript     ✓ ✓ ✓ ✓ ✓ 

  Legend: H[ighlight], L[ocals], F[olds], I[ndents], In[j]ections
         +) multiple parsers found, only one will be used
         x) errors found in the query, try to run :TSUpdate {lang}
```

### Output of `nvim --version`

```text
NVIM v0.6.1
Build type: Release
LuaJIT 2.1.0-beta3
Compiled by builduser

Features: +acl +iconv +tui
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/share/nvim"

Run :checkhealth for more info
```

**Comment by baltitenger** [comment]
@clason assuming it's enough to install the plugin and call `require'matchparen'.setup()`, it doesn't fix it.

**Comment by clason** [comment]
It should be, yes. That's good to know, and shows that `matchparen` is in fact not the culprit here.

**Comment by baltitenger** [comment]
Yeah, well I guess it has to do with `matchaddpos`, if I comment it out at the end of `s:Highlight_Matching_Pair` in `plugin/matchparen.vim`. Either way, this is probably not the main cause of the problem, since even if I disable `matchparen`, `:TSHighlightCapturesUnderCursor` says there aren't any highlight groups there, when I actually see it being highlighted.

**Comment by clason** [comment]
That means there are actually two different issues, then:
1. missing highlighting (with `matchaddpos`)
2. `TSHighlightCapturesUnderCursor` stopping at a certain level of recursion (which is probably an issue for playground, not here?)

**Comment by baltitenger** [comment]
What's weird is that it only happens with html. e.g. if I were to create a function nested inside functions more than 5 deep in javascript, the highlighting and `:TSHighlightCapturesUnderCursor` both work as expected.

Also, the playground itself does display the syntax tree correctly, and looking at `playground/lua/nvim-treesitter-playground/hl-info.lua`, I don't really see anything that should imply a recursion depth limit, seemingly it just uses neovim's builtin treesitter querying functions. (I don't know much about how this works so I might be totally wrong)

**Comment by baltitenger** [comment]
I think `Query:iter_captures` is doing something wrong with the bounds checking. If I change [`nvim-treesitter-playground/hl-info.lua:39`](https://github.com/nvim-treesitter/playground/blob/master/lua/nvim-treesitter-playground/hl-info.lua#L39) to omit the `start` and `stop` parameters from the `iter_captures` call, while it probably makes it really inefficient, `:TSHighlightCapturesUnderCursor` now correctly reports the highlight groups. Something similar might be happening in the highlighting part as well.
